semgrep-core/Core/AST/AST_generic.ml

```diff
@@ -1226,6 +1226,11 @@ and other_type_argument_operator =
   | OTA_Todo
   (*e: type [[AST_generic.other_type_argument_operator]] *)
 
+and tuple_type_member =
```

Comment:
hmm nah. You can use TyQuestion for TyTupOpt
and we can define a new constructor for TyRest.
No need to change all those xxx_to_generic.ml just because Typescript does thing differently.

Reply:
I didn't realize there was already a TyQuestion in the generic AST when I wrote my comment on the pfff pull request earlier. https://github.com/returntocorp/pfff/pull/394#discussion_r576309641

Let me summarize what needs to change:
* Remove the following from AST_generic, use plain `type_` for tuple members:
```
and tuple_type_member =
  | TyTupMember of type_  (* simple tuple type element *)
  | TyTupOpt of type_  (* optional tuple type element in typescript: string? *)
  | TyTupRest of type_  (* rest tuple type element in typescript: ...string *)
```
* Add `TyRest` to `type_` in the generic AST:
```
and type_ =
...
  | TyPointer of tok * type_
  | TyRef of tok * type_ (* C++/Rust *)

  | TyQuestion of type_ * tok (* a.k.a option type *)
  | TyRest of tok * type_ (* '...foo' e.g. in a typescript tuple type *)
  ^^^^^^^^^^^^^^^^^^^^^

  (* intersection types, used for Java Cast, and in Typescript *)
  | TyAnd of type_ * tok (* & *) * type_
...
```
* The order of tuple elements in the generic AST follows the concrete syntax, that's why we have `type_ * tok` for `foo?` but `tok * type_` for `...foo`.

Reply:
The mapping from JS to generic AST is:
```ocaml
and tuple_type_member x =
  match x with
  | TyTupMember x -> x
  | TyTupOpt (type_, x) -> TyQuestion (type_, x)
  | TyTupRest (x, type_) -> TyRest (x, type_)
```